protocols/plaintext/src/handshake.rs

```diff
@@ -0,0 +1,198 @@
+            Ok(prop) => prop,
+            Err(_) => {
+                debug!("failed to parse remote's exchange protobuf message");
+                return Err(PlainTextError::HandshakeParsingFailure);
+            },
+        };
+
+        let pb_pubkey = prop.take_pubkey();
+        let public_key = match PublicKey::from_protobuf_encoding(pb_pubkey.get_Data()) {
+            Ok(p) => p,
+            Err(_) => {
+                debug!("failed to parse remote's exchange's pubkey protobuf");
+                return Err(PlainTextError::HandshakeParsingFailure);
+            },
+        };
+        let peer_id = match PeerId::from_bytes(prop.take_id()) {
+            Ok(p) => p,
+            Err(_) => {
+                debug!("failed to parse remote's exchange's id protobuf");
+                return Err(PlainTextError::HandshakeParsingFailure);
+            },
+        };
+
+        Ok(HandShakeContext {
+            config: self.config,
+            state: Remote {
+                peer_id,
+                public_key,
+            }
+        })
+    }
+}
+
+pub fn handshake<S>(socket: S, config: PlainText2Config)
+    -> impl Future<Item = (Framed<S, BytesMut>, PublicKey), Error = PlainTextError>
+where
+    S: AsyncRead + AsyncWrite + Send,
+{
+    let socket = length_delimited::Builder::new()
+        .big_endian()
+        .length_field_length(4)
+        .new_framed(socket);
+
+    future::ok::<_, PlainTextError>(HandShakeContext::new(config))
+        .and_then(|context| {
+            trace!("starting handshake");
+            Ok(context.with_local()?)
+        })
+        // Send our local `Exchange`.
+        .and_then(|context| {
+            trace!("sending exchange to remote");
+            socket.send(BytesMut::from(context.state.exchange_bytes.clone()))
+                .from_err()
+                .map(|s| (s, context))
+        })
+        // Receive the remote's `Exchange`.
+        .and_then(move |(socket, context)| {
+            trace!("receiving the remote's exchange");
+            socket.into_future()
+                .map_err(|(e, _)| e.into())
+                .and_then(move |(prop_raw, socket)| {
+                    let context = match prop_raw {
+                        Some(p) => context.with_remote(p)?,
+                        None => {
+                            debug!("unexpected eof while waiting for remote's exchange");
+                            let err = IoError::new(IoErrorKind::BrokenPipe, "unexpected eof");
+                            return Err(err.into());
+                        }
+                    };
+
+                    trace!("received exchange from remote; pubkey = {:?}", context.state.public_key);
+                    Ok((socket, context))
+                })
+        })
+        // Check the validity of the remote's `Exchange`.
+        .and_then(|(socket, context)| {
+            let is_valid = match context.state.peer_id.is_public_key(&context.state.public_key) {
```

Comment:
It's not the role of the plaintext protocol to check whether the public key matches the expected `PeerId`. That's already handled in the general logic of libp2p-core::nodes.
We don't necessarily know what the `PeerId` of the remote is going to be when we initiate the connection.

Reply:
This (checking that the transmitted peer ID is "consistent with" the transmitted public key) seems to be required by the spec though:

> Upon receiving an Exchange message from the remote peer, each side will validate that the given peer id is consistent with the given public key by deriving a peer id from the key and asserting that it's a match with the id field in the Exchange message.

This is _in addition to_:

> Dialing a peer in libp2p requires knowledge of the listening peer's peer id. As a result, the dialing peer ALSO verifies that the peer id presented by the listening peer matches the peer id that they attempted to dial.

Reply:
Oh ok. I misunderstood the code and for some reason thought that `context.state` was the configuration.